### ISSUE TITLE
Please release the Android app under a free license

### ISSUE THREAD

**Issue by pizzamaker** [mapillary/mapillary_issues#1124]
The Windows Phone app is already FLOSS. It would be nice to have Mapillary in F-Droid :)

**Comment by de-vries** [comment]
> It would be nice to have Mapillary in F-Droid

:+1: I vote for this!

**Comment by jesolem** [comment]
We want this too and will get there.

**Comment by domdomegg** [comment]
:+1:

**Comment by yogiks** [comment]
:+1:

**Comment by shlublu** [comment]
:+1:

**Comment by mnalis** [comment]
I agree conpletely, having the app in f-droid is prerequisite for me to start using it!

**Comment by pizzamaker** [comment]
@jesolem  What's the hold-up at this point?

**Comment by jesolem** [comment]
@pizzamaker We're on it, but a LOT on the stack right now. Sorry we can't do this faster. Expect it out very soon, @peterneubauer is working on it.

**Comment by pizzamaker** [comment]
@peterneubauer Any ETAs? I really want to use this application :)

**Comment by kozross** [comment]
I *definitely* support the code for this being released under a free software license. Google can't be allowed to rule the world, and this being free software would make a very positive start to ending that dominance.

**Comment by svn-king** [comment]
this is a great idea, it's what's been keeping me from using this app

**Comment by icantsleep** [comment]
This would be a very positive step forward. Is there a roadmap for this?

**Comment by rm-rf-github** [comment]
Really hoping this comes through

**Comment by Tirifto** [comment]
I believe this would make a wonderful contribution to free software, and may help a lot of people. Please do free this program, it would really be great to have.

**Comment by sstangl** [comment]
:+1:, would love to use on free Cyanogen.

**Comment by PureTryOut** [comment]
Would love to see this on F-Droid!

**Comment by blobgo** [comment]
Yes your service is already about OpenData and so as much as possible should also open source. **Especially** apps which request location data, camera permissions and internet access at the same time. (I know these permissions are needed, but you could track users in a very good way)

**Comment by pizzamaker** [comment]
Also related:  #469

**Comment by redsteakraw** [comment]
FYI I stopped contributing to Mapillary because of the lack of an open source client on F-Droid.  I won't contribute until there is one.

**Comment by pizzamaker** [comment]
@redsteakraw Same. And I've been waiting on this one for over a year.

**Comment by aablaablaa** [comment]
I want that too! if it would be foss and on F-Droid I would be contributing a lot more! :)

**Comment by peterneubauer** [comment]
if you want to roll your own - there is all the code you need in JOSM already - oauth, upload and submission of changesets. https://github.com/JOSM/Mapillary/tree/master/src/org/openstreetmap/josm/plugins/mapillary/oauth

**Comment by hughdavenport** [comment]
Does that mean that mapillary are no longer intending to release their own official app as open source, and instead relying on community to provide third party open source apps?

From https://github.com/mapillary/mapillary_issues/issues/469#issuecomment-231137575 it sounded very much like you were only planning on refactoring the current app to make it cleaner before releasing, but your latest comment sounds like that wouldn't even be a priority. Myself (and others it sounds like with some comments) are confused as to what in the current app you don't want released under open source, as there isn't anything clearly business logic orientated that you may want to hide. Instead it is simply an app that allows taking photos attaching tags to photos based on location, and uploading those photos to a server with known api endpoints. Of those features i honestly don't see what would not be possible to just open source as is. Then the community can help improve the official app instead of making several redundant copies of what would essentially be the same app.

Its saddening that this bug has been open for over a year with seemingly no progression and now the company just suggests to make your own.

I think the project as a whole is a great idea, and I've contributed a bunch to it, but sadly I have stopped until there is a decent open source app released. I would love to be able to contribute to this app as well.

**Comment by hughdavenport** [comment]
No comment?

**Comment by jesolem** [comment]
@hughdavenport I just added an update to #469 . Pasting below as well. Thanks for your patience. 


**The plan** is as @peterneubauer describes in #469 . Refactor to make an SDK together with an open source app using the SDK. We are getting requests from app developers to create support for Mapillary integration in their apps (capture, login, upload) and the SDK + sample app seems to be the best way to make that happen. 

**The timeline** looks like late spring for Android (May?) but is a bit dependant on when we can onboard an Android developer (Richard is no longer with us and we are interviewing candidates now). For iOS we are looking at the same, possibly a little earlier.

**Comment by hughdavenport** [comment]
First, thank you for taking a whole two months to reply, it really shows you care.

From what i gather, you currently don't have enough android developers, you have failed to mention what in your app is so heavily based on hush hush secret business logic, and you are once again just giving a timeline of just a couple of months away (which you have done several times in the past).

Can i make a suggestion... Open source as is, show the community that you actually believe in ooen source and want to be part of the community, and you never know, maybe the community can help you split out the SDK for you...

**Comment by jesolem** [comment]
@hughdavenport There is no need to make sarcastic comments. We appreciate all input and feedback and we're working as fast as a team of our size can. The estimate above is the best we can do at this point.

I understand that this issue is important to you but please also understand that we have far more people asking for 360 camera support, action camera support, and real bugs on Android. Priorities.

**Comment by domdomegg** [comment]
More 360 support/"real bugs" on android could be fixed by releasing it under a free license allowing external contributors to fix them!

**Comment by rugk** [comment]
Indeed, opens-sourcing something is easy. It does not need much work. You can fix the bugs anyway afterwards…

**Comment by hughdavenport** [comment]
I feel that a bug that has had almost a hundred comments and has been open for two years is something worth prioritising and shows that your users want it.

Open sourcing is probably the most easiest thing, create new github repo, push code as is. You might be surprised how many people help out with these features.

Was sarcastic because you are not listening to my opinion, and many others, and you are calling this a "not real bug" which clearly shows that you aren't really prioritising it at all.

**Comment by domdomegg** [comment]
Ermm... Agree with sentiment but nearly 100 comments? I count 30

**Comment by hughdavenport** [comment]
Theres about 3 different bugs open for it. I'll find in a sec

**Comment by hughdavenport** [comment]
There is #1945 which has some responses, #469 (2 years old) which has some different repsonses, #1791 which also has some muttering about how they might open source, but unsure how, then this bug, which was mysteriously broken for almost a year which may just be github but a big coincidence. There is also #1416 which is talking about splitting out the app into components, again looks very stalled. Another comment not replied to on #1399 when someone asks about this split out / light version of the app.

So that was maybe 5 minutes search on mobile, it shows the company regularly just ignores comments on the tracker for months at a time, keeps on saying they are in the process of splitting it out, or making a light version, or saying they want to open source. Nothing happens, then several months later they say exactly the same thing with a similar timeframe of completion of "a couple of months".

One comment directed to me was suggesting that i just make my own open source app as all the code is available, so they seem to think its easy for a random off the internet, yet too hard for them, but also don't want to open source to get help from internet randoms.

My personal experience with this app has cost me hundreds of dollars in data usage due to " only upload on wifi" meaning completely thr opposite (#2361).

Might not be 100, but is definitely more than expected for a company that says it prides itself on being open

**Comment by mnalis** [comment]
@jesolem well, I'm willing to give you guys a benefit of the doubt for one more time. We'll see by midyear if it was warranted, I guess.

If Android (or even iPhone for which you seem to have more developers) app goes open source on github in some form (even if it is not fully working version) before fall, it will show the community that you try to take the issue seriously, but are hampered by lack of manpower.

But if even by the end of 2017 **still nothing** of smartphone source code materializes, but it remains just on promises, then I guess much more folks would start sharing @hughdavenport belief that it's being sabotaged on purpose - it really does not take that much time to type "git push" (nor to add README with few lines describing that is still poorly documented and not yet fully working source)

**Comment by jesolem** [comment]
@mnalis We appreciate you sticking with us. 

It was never the intention to make anyone disappointed. And we can definitely improve our communication around the topic, especially if things don't seem to develop as fast as both you and we would have hoped.

We're doing spring cleaning and putting new routines in place for avoiding issues "on hold" like this for long periods. We will get back to you with an update on the open sourcing as soon as we can.

**Comment by bobblkabb** [comment]
I join as soon as the app is available on f-droid.

**Comment by aligitor** [comment]
I try to use exclusively floss softwares, only 2 apps are not free on my phone - one of them is Mapillary. Although I use it, but I will recommend it to my friends only when it will be released under free license. I'd appreciate it very much!

**Comment by rugk** [comment]
I just noticed a horrible trust problem in their applications. Why on hell does the Android app need the contact permission? https://github.com/mapillary/mapillary_issues/issues/2689

In a closed-source application they cannot prove that they don't upload the address book and sell all the data. Is that how Mapillary makes money? Exploiting their contributors?

We **need** the source code of your app, **now**, and not at any point in the future! Even if there are closed-source dependencies then the community can find replacements, or there will be a fork with a completly free/libre version, whatever… we'll see. But **give us the source**! Otherwise neither the project, nor the company is trustworthy!

**Comment by HolgerJeromin** [comment]
@rugk
Nobody is (hopefully) forced to use an app and a service of a company you do not trust.
You are not in the position to demand stuff in this rude tone. 
The financing of mapillary is no secret (and is not selling metadata).

And please stop spamming this issue tracker with duplice content.

**Comment by rugk** [comment]
Yeah, as it looked the contact permission was a stupid display, so it is okay… Anyway, that does not change the fact, that I can **expect** (not demand, of course) such an app to be FLOSS.

And it is more than justified to change the tone when a company refuses to do so for more than **two years**!

**Comment by jidanni** [comment]
I'll notify our Taiwan group and see if they are still 
enthusiastic about the project.
 https://m.facebook.com/groups/1138239502943100

**Comment by bmassot** [comment]
👍

**Comment by hughdavenport** [comment]
@HolgerJeromin I think it is reasonable that people question Mapillaries motives to be honest. Correct me if I'm wrong, but this issue has been open for multiple *years* with multiple comments from ex and current users of the application stating displeasure on the timelines given. The responses from the company never quite say that this will ever be open source or not.

When it is pointed out that this ticket is so old and the contributors don't comment, we get a comment such as https://github.com/mapillary/mapillary_issues/issues/1124#issuecomment-289035813 from *months* ago. Can we have that promised update? Or have those routines put in place to avoid stale issues such as this not doing what those routines should be doing?

Going back further, with a timeline given of late May, in https://github.com/mapillary/mapillary_issues/issues/1124#issuecomment-285373283. It's past that now by several *months*. Can Mapilliary outline why this target wasn't met, and what they are doing to try and meet it?

Can we simply get a decent timeline on when you will be open sourcing, that you aim very hard to stick with. I think you should agree that this issue has been going on long enough and multiple people have shown displeasure. I'm sure you understand (as your company says it prides itself in being open source), that once you open source, you can get help from developers around the world to fix some of the bugs you have, and help implement new features. I understand that this may place a burden on your developers to triage and clean up any code that is contributed, but in the long run this will help your company build efficient projects easily with a good document on contribution guidelines. If you think this won't be the case, I would be more than pleased to get any links from you of open source projects that are failing due to the fact they are open source. I can certainly find many projects that have thrived, just by searching on github.

If it is simply that you don't want to open source, click the Close button, then the displeased people can leave for good. I personally hope that this isn't the option you choose as this is a great project that should be open to the community. There is already an alternative for viewing Mapillary in a open source app OsmAnd, which I would currently suggest to all the people in this thread that are unhappy with Mapillary's open source stance to start using now.

Cheers,

Hugh

**Comment by katrinhumal** [comment]
@hughdavenport Thanks for your input. We're not closing this issue (or the others) because, as you have pointed out, there are several people who are interested in this topic and want to chime in every once in a while (so they shouldn't need to open a new issue every time). Also, we couldn't even do it on the basis of "if you don't want to open source just close it" since it's not correct.

Open sourcing the Android app (since this discussion shouldn't be expanded to open sourcing all of Mapillary - there are parts that are already open source, as well as parts that likely never will be) has its pros and cons for us, and so far the weights have not been favourable according to our evaluation (which is clearly different from yours). A huge piece of this being, as said before, that we haven't managed to fix the app into a state that we're happy with. And that's also why we don't want to give any more timeframes right now. We are sorry to have promised and not delivered. Can't take that back anymore but that's just how it is.

**Comment by rugk** [comment]
First great that you still plan to open-source it.

> has its pros and cons for us, and so far the weights have not been favourable according to our evaluation

You never mentioned any cons, except of "we need to clean up code", which is not a con, as you can still open-source and clean up the code with the help of the community. Nobody will blame you for that, you will only be blamed if you do not release the source. (I don't know what you fear here…)

> we haven't managed to fix the app into a state that we're happy with

Seriously if you as the devs of the app don't like the app I would be very worried about it. If the app is really so terrible as you imply, you would need to have a bad vibe each time you release the app on Google Play or so.
But actually I don't believe your app is so bad. It just needs to compile, to work, etc. That's enough, then you can go ahead and open-source it.

So I do not need any timeframe for releasing the source, but actually the answer to the question: Why not release it right away? Just publish your git repo, create a new git repo or whatever. That's an easy step.

**Comment by Hapoofesgeli** [comment]
Mapillary, unlike how it tries to show itself is pretty much a completely closed ecosystem, which makes it even worse than actual closed ecosystems. The fact that they get your data, grant themselves a licence out of it, extract and sell whatever useful comes out of it while locking you away is basically parasitism, especially when they try to show it otherwise. And making the Android app open source won't simply change that fact, even worse, it will make them yet more pretentious. It's sad seeing Mapillary's footprints in many open ecosystems I tend to use, and I really recommend people thinking twice before they contribute to whomever tends to be the next Google.

**Comment by hughdavenport** [comment]
@katrinhumal ok, one question, from the *very first comment of this thread*

> The Windows Phone app is already FLOSS. It would be nice to have Mapillary in F-Droid :)

Has the windows app been changed at all due to the devs thinking anything is wrong with it *in the last 2 years*?

How different is it from the android app? (I haven't actually got a windows phone so can't see)

What was your thought process years back when you decided to open source the windows one but not the android. It seems a bit backward to me lol :p

Ok, that was a few questions, but all branched off the same thing. I think you could probably agree that they are valid. I still don't see the reason for holding back code just because you don't think it is ready, and as @rugk says, it makes it sound like your app is terrible and that *no-one* should install it because of that. Is that really the message you want to send people?

@Hapoofesgeli how is replicant going for you ;)? Closed ecosystems are part of the technology world, and infact can help bring technology to the masses, and bring in more contributors to open source systems. Mapillary have done a stunning job IMHO, bringing images from around the world into one app. Some countries wouldn't give a flying **** whether it is open or not, more whether it *exists*. The fact that you use github makes me think hmm, you know github is closed? Might be running on open source rails, but the code is closed, they have TOS that limit you, but the *idea* that they brought around is amazing, and has improved the open source world because of it, and now there are several competitors for github that are completely open. Android, open source? at its base yes, in general no, most of the /drivers/, /additional apps/ etc bundled with any standard phone are not open source. cyanogen and forks helped that somewhat by changing the bundled apps, but the drivers are still not. What phone do you use? Are the drivers open source? Same goes for basically any linux laptop, are all your drivers open? Everything is going to have some closed aspect of it, or have performance/compatability issues if completely open until the technology comes up to standard. Sure I'm all for open source, but I think it is wrong to be completely against closed source if you want to contribute in this industry. I think you should think twice before jumping on your internet keyboard and slamming down a perfectly respectable company about their ecosystem and their product just because of your feelings about the world. I don't personally see how that remark would help this issue at all. If you don't think it is good then don't use it, don't try and make others see your bigotted opinion.

**Comment by Hapoofesgeli** [comment]
@hughdavenport It's funny how you simply interpret "closed" as if it could only mean "closed source". Github is sure closed source, and I try to avoid it as far as possible when there is not much to gain from it (the fact that I have no working repositories here should simply show that), but, there are many small and big differences between Github and Mapillary, I'll just point one simple one:

**Github:**
> The licenses you grant to us will end when you remove Your Content from our servers

**Mapillary:**
> Contributors grant Mapillary a worldwide, **perpetual**, and transferable license to use the images and other content

So yeah before you start attacking me for no reason and trying to teach me what is open source and what is not, you might want to notice some more stuff. Like the fact that Openstreetcam existed from 5 years before Mapiilary. Note that this is not to discuss why it failed and which is better etc, just to show you that unlike what you said, many open alternatives exist before the closed ones appear.

"I think it is wrong to be completely against closed source"
Although I am against  it at any case, that has little to to with what I said, again you have failed to see the difference between data and source code.

And as your favourite Github allows me, I'm allowed to say whatever I like, till I can't. So how about you bring your complaints  to it, rather than personally attacking me to defend something that barely has a value to you. And as "someone is wrong on the internet" discussions don't reach anywhere there is no point in continuing this any further, so good luck.

**Comment by hughdavenport** [comment]
... You simply have had no conversations in this ticket apart from to attack mapillary, not sure how that is helpful for you or anyone actually involved. I understand that there is a closed data issue, but as there isn't really a viable alternative that isn't then I don't personally see an issue with it, and any step towards being more open helps alternatives come. Your comments simply don't help that cause.

github was merely just one example, you can bring up many examples of popular open source projects with closed data, but lets not go there.

I also understand that openstreetcam existed (was openstreetview back then), but also understand that it actually gained a lot more traction after mapillary launched, hard to know reasons why though, but may be due to people wanting more open source/data/whatever. I think we should also note that google street view was around from even before then.

As for "someone is wrong on the internet", not sure what your original comment was, but it sure seemed like you were pointing out that mapillary, and everyone on this ticket were wrong, which I found quite attacking personally as I believe that I'm working towards a better system which is more open. Might not be 100%, or even 1% open in terms of data, but I think it is still a worthy project to follow and didn't appreciate you jumping in after not having participated ever before and saying that I'm wrong to think that. I'm sorry I might have offended you in my response, understand that I didn't mean to come across as attacking but I typed that in the heat of the moment after reading your comments.

**Comment by Alessandro-Barbieri** [comment]
Mapillary and openstreetcam have different goals

**Comment by rugk** [comment]
Eh, really which? See also the issue in OpenStretCam where I asked about differences/collaboration and after a long discussion i think we know there were not much differences, from a license perspective e.g.

**Comment by Alessandro-Barbieri** [comment]
@rugk openstreetcam goal is to collect photos for improving openstreetmap

**Comment by rugk** [comment]
Mappilaries goal is alsop to collect photos, e.g. for improving OpenStreetMap.

**Comment by HolgerJeromin** [comment]
Windows phone app was not build by mapillary staff but one single private dev. Thats why it has a different licence.

**Comment by jaakkoh** [comment]
@rugk , Regarding differences between Mapillary and OpenStreetCam it's noteworthy to remember that OpenStreetCam is a project fully financed and run by Telenav, Inc., a publicly traded company (NASDAQ:TNAV) with its core business in wireless location-based services including satellite navigation, local search, automotive navigation solutions, mobile advertising, enterprise mobility and workflow automation (per Wikipedia).

With the above it is not a surprise that OpenStreetCam has been developed rather purely to capture roads. By opening the app it is (or at least was a year ago) clear from the very beginning.

I'm not saying this is a bad thing by itself. But I think it's worth keeping in mind that the owner of the project guides the development of the project heavily (as a common note) and I find this to be the case with OpenStreetCam too. I'd also note that while OpenStreetCam is positioned as and technically released to be open _source_ (I don't know if totally fully -- are there e.g. full dumps of the imagery available?) it's just one aspect in the big picture.

**Comment by rugk** [comment]
> (I don't know if totally fully -- are there e.g. full dumps of the imagery available?)

Not even the app is fully open-source as you can see it has the same problem as Mapillary here. It is also still not included in F-Droid:

https://github.com/openstreetcam/android/issues/8

**Comment by jaakkoh** [comment]
@rugk Hmm. I didn't know that and am slightly surprised. Then again, maybe not really.

**Comment by ltog** [comment]
New year, new hope...

How is the FLOSS client coming?

**Comment by peterneubauer** [comment]
Actually, we have internally started on the iOS side to break out the
essential functions and put them into an Sdk. Can't give an exact public
release date yet, but yes, it's under way. This issue is by no means
forgotten or ignored - we just don't want it to dominate our roadmap.

**Comment by rugk** [comment]
Sure, just don't let free software "dominate your roadmap". Who needs that anyway…

Also just to be sure: we do not (only) want a floss SDK. Open source the whole app, so we can have it on F-Droid (as for the android app), etc.

**Comment by elgaard** [comment]
OpenStreetCam is not on F-droid. But it should be possible to fork it and make a free version.
With OpenStreetCam you can capture images (you have to disable MP4) without logging in. The images can then be accessed from a computer and uploaded to Mapillary. I made a small POC script to set the 
GPS EXIF from OpenStreetCam before uploading to Mapillary. https://github.com/elgaard/OpenStreetmapTools/blob/master/setPosFromTrack.py

This way I can capture images and upload them to both OpenStreetCam and Mapillary.
Of course, it would be better to be able to do this from the android app.
It would be great to change the OpenStreetCam app to capture images, and then prompting to upload them to OpenStreetCam, Mapillary, user-defined accounts (webdav, ssh etc.)

**Comment by alexgleason** [comment]
Interesting. I found Mapillary because OsmAnd~ recommends it through their app. Was surprised it's proprietary. Just wanted to echo the sentiment of others; I'd love to contribute, but I won't until the Android app is free.

It seems clear the authors are well intentioned. Still, I do not understand why improving the codebase should precede an Open Source release. I would appreciate if someone would be willing to elaborate on that.

Is it because of fear? Pride? Maybe the app contains bugs that others will see and exploit. Or maybe it could hamper your reputation because the code isn't good enough? Or maybe it will disappoint people. These feelings are common when taking a mature proprietary project and throwing its code into the wild. But I don't think breaking out the SDK will actually reduce these fears. It seems like a rationalization for avoiding this perceived risk.

Honestly, what is really the worst thing that could happen? Imagine a real play-by-play and ask yourself if it's realistic. If it's not, try again. Once you find some realistic bad outcomes, ask yourself if it's really that bad. Usually it's like a pinprick, a slight momentary pain, but with long-term benefits.

If I were you, I would really just go for it. It seems scary, but probably, everything will be just fine. There's a mantra in software development, "Fuck It, Ship It" that I really like. Like an artist who never sees their painting as truly "finished," you won't either, so you might as well push it out now, bearing the risk, before it's too late.

**Comment by utack** [comment]
I went ahead and created you your own personal newsletter  
![newsletter](https://user-images.githubusercontent.com/8551234/39275879-391f9f0c-48e7-11e8-80ca-8c726cb0e312.png)

**Comment by HolgerJeromin** [comment]
Just a comment from a person who works on a closed source software for living.

Releasing the code without allowing/supporting PRs from others is useless.
And managing and coordinating work from others also needs much work.
The author of streetcomplete for example investes a big amount of time reviewing and optimizing 3rd party code instead of coding.
Typescript as a second example has many pr open without reactions of the maintainers.

features can be out of scope for mapillary but they have to deal with the "noise".

So releasing the code is no "others will improve the app for free without costs" solution.

**Comment by rugk** [comment]
1. The already have much of the noise, because of thius issue tracker here. Most of the noise is froim issues anyway, not from PRs.
2. They can release the code without accepting PRs. Not nice, but totally possible – and certainly better than not releasing it at all.

**Comment by mnalis** [comment]
@HolgerJeromin I strongly disagree that releasing the code without supporting PRs is useless. Among other things, it allows for for forks which require no time at all from Mapillary (if they don't want it). Also, it allows for code inclusion in other products (for example, OsmAnd support is one I want for example - there is no way I'll use closed source app to contribute, and taking pictures and uploading manually is way too much volunteer work) 

If you ask yourself "why would they want others to use use code and not use any code from community back", then remember that Mapillary does not make a living by selling this software, but by getting volunteer produced data - and anything that gives them more exposure and more volunteers is good thing. 

Also, PR from others could be accepted with almost no work at all into separate "community" branch. Which not only makes "open source loving" people happy and contributing, but also gives additional benefit as Mapillary itself can cherry pick stuff they want. While importing to "official" branch will of course take some work, it is waaaay less work for Mapillary then coding/fixing those issues themselves. And of course, they can decide not to import (some or all) PRs to official branch if they so choose, it is also fine.

**Comment by alex9099** [comment]
It would be pretty nice to have the app open source, i only use F-Droid and compile apps from source, rarelly i touch play store

Currently i use OpenStreetCam because it is open source, i would consider to contribute to both, but closed source is a no go for me
